src/objects.c

```diff
@@ -1555,53 +1555,263 @@ const char *p11prov_obj_get_ec_group_name(P11PROV_OBJ *obj)
+    }
+    export_ctx->allocs[export_ctx->ialloc++] = p;
+    return p;
+}
+
+static void obj_export_ec_ctx_free(struct obj_export_ec_ctx *export_ctx)
+{
+    BN_CTX_end(export_ctx->bnctx);
+    BN_CTX_free(export_ctx->bnctx);
+    for (int i = 0; i < export_ctx->ialloc; i++) {
+        OPENSSL_free(export_ctx->allocs[i]);
+    }
+}
+
+static int ossl_param_construct_bn(OSSL_PARAM *param,
+                                   struct obj_export_ec_ctx *export_ctx,
+                                   const char *key, const BIGNUM *val)
+{
+    size_t bsize;
+    void *buf;
+
+    if (BN_is_negative(val)) {
+        return 0;
+    }
+
+    bsize = (size_t)BN_num_bytes(val);
+    /* We make sure that at least one byte is used, so zero is properly set */
+    if (bsize == 0) {
+        bsize++;
+    }
+
+    buf = obj_export_ec_malloc(export_ctx, bsize);
+    if (buf == NULL) {
+        return 0;
+    }
+
+    if (BN_bn2nativepad(val, buf, bsize) < 0) {
+        return 0;
+    }
+
+    *param = OSSL_PARAM_construct_BN(key, buf, bsize);
+    return 1;
+}
+
+static int ec_group_explicit_to_params(P11PROV_OBJ *obj, const EC_GROUP *group,
+                                       OSSL_PARAM *params, int *nparam,
+                                       struct obj_export_ec_ctx *export_ctx)
+{
+    int fid;
+    const char *field_type;
+    BIGNUM *p, *a, *b;
+    const BIGNUM *order, *cofactor;
+    const EC_POINT *generator;
+    point_conversion_form_t genform;
+    size_t bsize;
+    void *buf;
+    unsigned char *seed;
+    size_t seed_len;
+
+    fid = EC_GROUP_get_field_type(group);
+    if (fid == NID_X9_62_prime_field) {
+        field_type = SN_X9_62_prime_field;
+    } else if (fid == NID_X9_62_characteristic_two_field) {
+        field_type = SN_X9_62_characteristic_two_field;
+    } else {
+        P11PROV_raise(obj->ctx, CKR_GENERAL_ERROR, "Invalid EC field");
+        return RET_OSSL_ERR;
+    }
+
+    params[(*nparam)++] = OSSL_PARAM_construct_utf8_string(
+        OSSL_PKEY_PARAM_EC_FIELD_TYPE, (char *)field_type, 0);
+
+    export_ctx->bnctx = BN_CTX_new_ex(p11prov_ctx_get_libctx(obj->ctx));
+    if (export_ctx->bnctx == NULL) {
+        return RET_OSSL_ERR;
+    }
+    BN_CTX_start(export_ctx->bnctx);
+
+    p = BN_CTX_get(export_ctx->bnctx);
+    a = BN_CTX_get(export_ctx->bnctx);
+    b = BN_CTX_get(export_ctx->bnctx);
+    if (b == NULL) {
+        return RET_OSSL_ERR;
+    }
+
+    if (!EC_GROUP_get_curve(group, p, a, b, export_ctx->bnctx)) {
+        P11PROV_raise(obj->ctx, CKR_GENERAL_ERROR, "Invalid curve");
+        return RET_OSSL_ERR;
+    }
+
+    if (!ossl_param_construct_bn(&params[(*nparam)++], export_ctx,
+                                 OSSL_PKEY_PARAM_EC_P, p)
+        || !ossl_param_construct_bn(&params[(*nparam)++], export_ctx,
+                                    OSSL_PKEY_PARAM_EC_A, a)
+        || !ossl_param_construct_bn(&params[(*nparam)++], export_ctx,
+                                    OSSL_PKEY_PARAM_EC_B, b)) {
```

Comment:
please P11PROV_raise here as well

Reply:
or better do it in ossl_param_construct_bn() and mention for what key the operation failed.

Reply:
Done.